adapter/avro/src/main/java/org/apache/arrow/adapter/avro/producers/logical/AvroDecimalProducer.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.arrow.adapter.avro.producers.logical;
+
+import java.io.IOException;
+import org.apache.arrow.adapter.avro.producers.BaseAvroProducer;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.DecimalVector;
+import org.apache.avro.io.Encoder;
+
+/**
+ * Producer that produces decimal values from a {@link DecimalVector}, writes data to an Avro
+ * encoder.
+ */
+public abstract class AvroDecimalProducer extends BaseAvroProducer<DecimalVector> {
+
+  /** Instantiate an AvroDecimalProducer. */
+  protected AvroDecimalProducer(DecimalVector vector) {
+    super(vector);
+  }
+
+  /** Producer for decimal logical type with original bytes type. */
+  public static class BytesDecimalProducer extends AvroDecimalProducer {
+
+    private final byte[] reuseBytes;
+
+    /** Instantiate a BytesDecimalConsumer. */
+    public BytesDecimalProducer(DecimalVector vector) {
+      super(vector);
+      Preconditions.checkArgument(
+          vector.getTypeWidth() <= 16, "Decimal bytes length should <= 16.");
+      reuseBytes = new byte[vector.getTypeWidth()];
```

Comment:
Is this correct? I don't think the type width is constant for each value. It might be similar to what you did in `AvroBytesProducer`.

Reply:
Decimal is fixed width byte (either 16 or 32)

Reply:
Well, I think I have mixed up the requirement from Iceberg which says `Stored as fixed using the minimum number of bytes for the given precision.`

Reply:
Apologies, have removed this, I was mirroring consumers which need to handle both bytes and fixed inputs, but for producers we can always used fixed and the type width is always known.